src/proof_rules/mod.rs

```diff
@@ -399,3 +423,145 @@ impl<'tcx, 'sunit> VisitorMut for EncodingVisitor<'tcx, 'sunit> {
         Ok(())
     }
 }
+
+struct CallGraphVisitor<'tcx> {
+    tcx: &'tcx mut TyCtx,
+    current_proc: Option<Ident>,
+    graph: petgraph::graph::DiGraph<Ident, Span>,
+    node_indices: std::collections::HashMap<Ident, petgraph::graph::NodeIndex>, // map of proc name to node index
+}
+
+impl<'tcx> CallGraphVisitor<'tcx> {
+    pub fn new(tcx: &'tcx mut TyCtx) -> Self {
+        CallGraphVisitor {
+            tcx,
+            current_proc: None,
+            graph: petgraph::graph::DiGraph::new(),
+            node_indices: std::collections::HashMap::new(),
+        }
+    }
+}
+
+impl VisitorMut for CallGraphVisitor<'_> {
+    type Err = ();
+
+    fn visit_proc(&mut self, proc_ref: &mut DeclRef<ProcDecl>) -> Result<(), Self::Err> {
+        let proc_ref = proc_ref.clone(); // lose the reference to &mut self
+        let proc = proc_ref.borrow();
+
+        self.current_proc = Some(proc.name);
+
+        // Visit the body of the procedure
+        let mut body = proc.body.borrow_mut();
+        if let Some(ref mut block) = &mut *body {
+            self.visit_block(block)?;
+        }
+
+        self.current_proc = None;
```

Comment:
Probably unlikely to happen, but if the call `self.visit_block` fails, then we'll be in an incorrect state, no? One possibility is to handle the error from `self.visit_block` only after the `self.current_proc = None` assignment.